public/apps/configuration/panels/role-edit/tenant-panel.test.tsx

```diff
@@ -0,0 +1,142 @@
+
+describe('Role edit - tenant panel', () => {
+  const tenantName1 = 'tenant1';
+  it('buildTenantPermissionState', () => {
+    const input: RoleTenantPermission[] = [
+      {
+        tenant_patterns: [tenantName1],
+        allowed_actions: [TENANT_READ_PERMISSION],
+      },
+    ];
+
+    const result = buildTenantPermissionState(input);
+
+    const expected: RoleTenantPermissionStateClass[] = [
+      {
+        tenantPatterns: [{ label: tenantName1 }],
+        permissionType: TenantPermissionType.Read,
+      },
+    ];
+    expect(result).toEqual(expected);
+  });
+
+  it('unbuildTenantPermissionState', () => {
+    const permissions: RoleTenantPermissionStateClass[] = [
+      {
+        tenantPatterns: [{ label: tenantName1 }],
+        permissionType: TenantPermissionType.Full,
+      },
+    ];
+
+    const result = unbuildTenantPermissionState(permissions);
+
+    const expected: RoleTenantPermission[] = [
+      {
+        tenant_patterns: [tenantName1],
+        allowed_actions: [TENANT_READ_PERMISSION, TENANT_WRITE_PERMISSION],
+      },
+    ];
+
+    expect(result).toEqual(expected);
+  });
+
+  describe('TenantPanel', () => {
+    const tenantName2 = 'tenant2';
+    const optionUniverse = [{ label: tenantName1 }, { label: tenantName2 }];
+    const setState = jest.fn();
+
+    it('render an empty row if data is empty', () => {
+      shallow(<TenantPanel state={[]} optionUniverse={optionUniverse} setState={setState} />);
+
+      expect(setState).toHaveBeenCalledWith([
+        {
+          tenantPatterns: [],
+          permissionType: TenantPermissionType.Full,
+        },
+      ]);
+    });
+
+    it('render data', () => {
+      const state: RoleTenantPermissionStateClass[] = [
+        {
+          tenantPatterns: [{ label: tenantName1 }],
+          permissionType: TenantPermissionType.Full,
+        },
+        {
+          tenantPatterns: [{ label: tenantName2 }],
+          permissionType: TenantPermissionType.Read,
+        },
+      ];
+
+      const component = shallow(
+        <TenantPanel state={state} optionUniverse={optionUniverse} setState={setState} />
+      );
+
+      const comboBoxArray = component.find(EuiComboBox);
+      expect(comboBoxArray.length).toEqual(2);
+      expect(comboBoxArray.at(0).prop('selectedOptions')).toBe(state[0].tenantPatterns);
+      expect(comboBoxArray.at(1).prop('selectedOptions')).toBe(state[1].tenantPatterns);
+
+      const superSelectArray = component.find(EuiSuperSelect);
+      expect(superSelectArray.at(0).prop('valueOfSelected')).toBe(state[0].permissionType);
+      expect(superSelectArray.at(1).prop('valueOfSelected')).toBe(state[1].permissionType);
+    });
+
+    it('add row', () => {
+      const component = shallow(
+        <TenantPanel state={[]} optionUniverse={optionUniverse} setState={setState} />
+      );
+
+      const addRowButton = component.find(EuiButton).last();
+      addRowButton.simulate('click');
```

Comment:
nonblocking: we are using first and last since there is no id identifier for the buttons. However, the reader may need to check the UI to understand it. One minor improvement is to add one more expect about the button text.